### PR TITLE
Sealed Array Implementation

### PR DESCRIPTION
**Pull request by Anoukh**
## Purpose
This PR will introduce the concept of a sealed array, where any array type can be sealed in respect to the number of elements it can hold

## Goals
Provide a maximum size for arrays being declared

#### To a seal an integer array to a size of 5:
``` ballerina
int[5] sealedArray1 = [2, 15, 200, 1500, 5000];
int[5] sealedArray2; // Integer array of size five, filled with default integer values
```

#### To a seal an array by inferring the size from the right hand side expression, `sealed` keyword is used:
``` ballerina
sealed int[] sealedArray1 = [2, 15, 200, 1500, 5000]; // Create a sealed array of size 5
```

### CODE REVIEW

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/types/BArrayType.java [hunk 1]
```diff
@@ -39,6 +39,7 @@
 public class BArrayType extends BType implements BIndexedType {
     private BType elementType;
     private int dimensions = 1;
+    private int size = -1;
```
Comment (hasithaa) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/types/BArrayType.java]: How we handle multi-dimension sealed array? I think we need to store size array here.

Comment (Anoukh) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/model/types/BArrayType.java]: In a multi-dimensional array, elementType will be an array with a size if that array is also sealed

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/types/BArrayType.java [hunk 2]
```diff
@@ -48,13 +49,47 @@ public BArrayType(BType elementType) {
...
+            switch (tag) {
+                case TypeTags.INT_TAG:
+                    return (V) new BIntArray(size);
+                case TypeTags.FLOAT_TAG:
+                    return (V) new BFloatArray(size);
+                case TypeTags.BOOLEAN_TAG:
+                    return (V) new BBooleanArray(size);
+                case TypeTags.STRING_TAG:
+                    return (V) new BStringArray(size);
+                case TypeTags.BLOB_TAG:
```
Comment (hasithaa) [inline bvm/ballerina-core/src/main/java/org/ballerinalang/model/types/BArrayType.java]: Need to update with byte type.

Comment (Anoukh) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/model/types/BArrayType.java]: Fixed in commit d2ba493

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java [hunk 3]
```diff
@@ -667,7 +678,19 @@ public void visit(BLangJSONArrayLiteral arrayLiteral) {
...
+
+        BLangLiteral sealedSizeLiteral = new BLangLiteral();
+        sealedSizeLiteral.pos = arrayLiteral.pos;
+        sealedSizeLiteral.value = -1L;
+        if (arrayLiteral.type.tag == TypeTags.ARRAY &&
+                ((BArrayType) arrayLiteral.type).state != BArrayState.UNSEALED) {
+            sealedSizeLiteral.value = (long) ((BArrayType) arrayLiteral.type).size;
+        }
+        sealedSizeLiteral.type = symTable.intType;
+        genNode(sealedSizeLiteral, this.env);
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java]: Shall we move this common logic (writing the integer literal part) to a helper method? there are three places the same logic is used, so we can replace them..

Comment (Anoukh) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/codegen/CodeGenerator.java]: Addressed in 2e06894

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/types/BArrayType.java [hunk 4]
```diff
@@ -95,6 +130,9 @@ public int hashCode() {
     public boolean equals(Object obj) {
         if (obj instanceof BArrayType) {
             BArrayType other = (BArrayType) obj;
+            if (other.size != -1 && this.size != other.size) {
```
[the remark this answers is not in the document]
Comment (Anoukh) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/model/types/BArrayType.java]: Addressed in 2e06894

File: bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BRefValueArray.java [hunk 5]
```diff
@@ -28,20 +29,27 @@
...
     }
 
     public BRefValueArray(BType type) {
-        this.arrayType = type;
-        values = (BRefType[]) newArrayInstance(BRefType.class);
-        Arrays.fill(values, type.getEmptyValue());
+        super.arrayType = type;
+        if (type.getTag() == TypeTags.ARRAY_TAG) {
+            BArrayType arrayType = (BArrayType) type;
+            if (arrayType.getSize() != -1) {
```
[the remark this answers is not in the document]
Comment (Anoukh) [reply bvm/ballerina-core/src/main/java/org/ballerinalang/model/values/BRefValueArray.java]: Addressed in 2e06894

**Pull request merged**